File: servidor.py
import socket
import threading
import random

clientes = {}
pontuacoes = {}
jogo_ativo = True

SERVER_IP = '127.0.0.1'
PORTA = 9999

servidor = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
servidor.bind((SERVER_IP, PORTA))

print(f"[SERVIDOR INICIADO] Escutando em {SERVER_IP}:{PORTA}")

def enviar(msg, endereco):
    servidor.sendto(msg.encode(), endereco)

def sortear_carta():
    return random.randint(1, 11)

def distribuir_carta(endereco):
    carta = sortear_carta()
    pontuacoes[endereco] += carta
    enviar(f"CARTA:{carta}", endereco)
    if pontuacoes[endereco] > 21:
        enviar("RESULTADO:perdeu", endereco)

def avaliar_resultado():
    vencedor = None
    maior = 0
    for end, pontos in pontuacoes.items():
        if pontos <= 21 and pontos > maior:
            maior = pontos
            vencedor = end

    for end in pontuacoes:
        if end == vencedor:
            enviar("RESULTADO:ganhou", end)
        else:
            enviar("RESULTADO:perdeu", end)

def ouvir():
    global jogo_ativo
    while True:
        msg, endereco = servidor.recvfrom(1024)
        msg = msg.decode()

        if msg.startswith("ENTRAR:"):
            nome = msg.split(":")[1]
            clientes[endereco] = nome
            pontuacoes[endereco] = 0
            print(f"[{nome}] entrou no jogo.")
            enviar("MENSAGEM:Bem-vindo ao jogo!", endereco)

            if len(clientes) >= 2:
                for end in clientes:
                    distribuir_carta(end)

        elif msg == "PEDIR_CARTA":
            if endereco in pontuacoes:
                distribuir_carta(endereco)

        elif msg == "PARAR":
            pontuacoes[endereco] += 0
            enviar("MENSAGEM:Jogador parou. Aguardando os outros...", endereco)

            if all(p > 21 for p in pontuacoes.values()):
                avaliar_resultado()

threading.Thread(target=ouvir).start()

File: servidor.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;

class Servidor
{
    static Dictionary<IPEndPoint, int> pontuacoes = new();
    static Dictionary<IPEndPoint, string> jogadores = new();
    static UdpClient udp = new(9999);
    static Random random = new();

    static void Main()
    {
        Console.WriteLine("[SERVIDOR INICIADO] Porta 9999");
        new Thread(() => Ouvir()).Start();
    }

    static void Ouvir()
    {
        while (true)
        {
            IPEndPoint origem = new(IPAddress.Any, 0);
            byte[] dados = udp.Receive(ref origem);
            string msg = Encoding.UTF8.GetString(dados);

            if (msg.StartsWith("ENTRAR:"))
            {
                string nome = msg.Split(":")[1];
                jogadores[origem] = nome;
                pontuacoes[origem] = 0;
                Console.WriteLine($"[{nome}] entrou no jogo.");
                Enviar("MENSAGEM:Bem-vindo ao jogo!", origem);

                if (jogadores.Count >= 2)
                {
                    foreach (var jogador in jogadores.Keys)
                        DistribuirCarta(jogador);
                }
            }
            else if (msg == "PEDIR_CARTA")
            {
                if (pontuacoes.ContainsKey(origem))
                    DistribuirCarta(origem);
            }
            else if (msg == "PARAR")
            {
                Enviar("MENSAGEM:Jogador parou. Aguardando os outros...", origem);
                VerificarFinal();
            }
        }
    }

    static void DistribuirCarta(IPEndPoint jogador)
    {
        int carta = random.Next(1, 12);
        pontuacoes[jogador] += carta;
        Enviar($"CARTA:{carta}", jogador);

        if (pontuacoes[jogador] > 21)
        {
            Enviar("RESULTADO:perdeu", jogador);
            VerificarFinal();
        }
    }

    static void VerificarFinal()
    {
        if (TodosFinalizaram())
        {
            int maior = 0;
            IPEndPoint vencedor = null;

            foreach (var kv in pontuacoes)
            {
                if (kv.Value <= 21 && kv.Value > maior)
                {
                    maior = kv.Value;
                    vencedor = kv.Key;
                }
            }

            foreach (var jogador in pontuacoes.Keys)
            {
                string resultado = jogador.Equals(vencedor) ? "ganhou" : "perdeu";
                Enviar($"RESULTADO:{resultado}", jogador);
            }
        }
    }

    static bool TodosFinalizaram()
    {
        foreach (var pontos in pontuacoes.Values)
        {
            if (pontos <= 21)
                return false;
        }
        return true;
    }

    static void Enviar(string msg, IPEndPoint destino)
    {
        byte[] dados = Encoding.UTF8.GetBytes(msg);
        udp.Send(dados, dados.Length, destino);
    }
}
